File: src/main/java/aunmag/shooter/game/environment/actor/Stamina.java
package aunmag.shooter.game.environment.actor;

import aunmag.shooter.core.utilities.UtilsMath;

public class Stamina {

    private static final float MIN = 0.01f;
    private static final float MAX = 1;
    private static final float RECURRENCE_STEP = 0.07f;

    private final Actor actor;
    private float stamina = MAX;

    public Stamina(Actor actor) {
        this.actor = actor;
    }

    public void update() {
        if (stamina < MAX) {
            stamina += RECURRENCE_STEP
                    * actor.world.time.getDelta()
                    * actor.getHealth();
        }
    }

    public void spend(float stepFactor) {
        var delta = (float) actor.world.time.getDelta();

        stamina = UtilsMath.limit(
                stamina - RECURRENCE_STEP * stepFactor * delta,
                MIN,
                MAX
        );
    }

    public float get() {
        return stamina;
    }

    public float getEfficiency() {
        return MAX - (float) Math.pow(MAX - stamina, 2);
    }

}

File: src/main/java/aunmag/shooter/game/environment/actor/Stamina.kt
package aunmag.shooter.game.environment.actor

import aunmag.shooter.core.utilities.UtilsMath

class Stamina(private val actor: Actor) {

    var current = 1.0f
        private set(value) {
            field = UtilsMath.limitNumber(value, reserve, 1.0f)
        }
    private val reserve = 0.01f
    private val recurrenceStep = 0.07f

    fun update() {
        if (current < 1.0f) {
            current += recurrenceStep * actor.world.time.delta.toFloat() * actor.health
        }
    }

    fun spend(stepFactor: Float) {
        current -= recurrenceStep * stepFactor * actor.world.time.delta.toFloat()
    }

    /**
     * ~/src/python/stamina_efficiency.py
     */
    fun calculateEfficiency(): Float {
        return 1.0f - Math.pow(1.0 - current, 2.0).toFloat()
    }

}
